### ISSUE TITLE
combine multimeasure expression language and multientity

### ISSUE THREAD

**Issue by AlvaroVega** [telefonicaid/iotagent-node-lib#941]
For a device provisioned like:
```json
{
 "devices": [
 {
 "device_id": "testmulti",
 "protocol": "IoTA-JSON",
 "expressionLanguage": "jexl",
 "attributes": [
 {
 "object_id": "cont1",
 "name": "vol",
 "type": "string",
 "expression" : "vol*100",
 "entity_name": "WaterMeterSoria01",
 "entity_type": "WaterMeter"
 },
 {
 "object_id": "cont2",
 "name": "vol",
 "type": "string",
 "expression" : "vol*100",
 "entity_name": "WaterMeterSoria02",
 "entity_type": "WaterMeter"
 },
 {
 "object_id": "cont3",
 "name": "vol",
 "type": "string",
 "expression" : "vol*100",
 "entity_name": "WaterMeterSoria03",
 "entity_type": "WaterMeter"
 }
 ]
 }
 ]
}
```

and a measure like:
$ mosquitto_pub -h localhost -t /json/apikey/testmulti/attrs -m '{"cont1": 4, "cont2": 5, "cont3":6}' 

```
time=2020-11-23T11:05:34.111Z | lvl=INFO | corr=addfa3c0-326c-46e7-815e-d0b6f98695d2 | trans=addfa3c0-326c-46e7-815e-d0b6f98695d2 | op=IoTAgentNGSI.NGSIService | from=n/a | srv=smartcity | subsrv=/ | msg=Using the following request:

{
    "url": "http://iot-orion:1026/v2/op/update",
    "method": "POST",
    "headers": {
        "fiware-service": "smartcity",
        "fiware-servicepath": "/"
    },
    "json": {
        "actionType": "append",
        "entities": [
            {
                "id": "thing:testmulti",
                "type": "thing"
            },
            {
                "id": "WaterMeterSoria01",
                "type": "WaterMeter",
                "vol": {
                    "type": "string",
                    "value": "400",
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-23T11:05:34.110Z"
                        }
                    }
                },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-23T11:05:34.110Z"
                }
            },
            {
                "id": "WaterMeterSoria02",
                "type": "WaterMeter",
                "vol": {
                    "type": "string",
                    "value": "400",
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-23T11:05:34.110Z"
                        }
                    }
                 },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-23T11:05:34.110Z"
                }
            },
            {
                "id": "WaterMeterSoria03",
                "type": "WaterMeter",
                "vol": {
                    "type": "string",
                    "value": "400",
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-23T11:05:34.110Z"
                        }
                    }
                 },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-23T11:05:34.110Z"
                }

            }
        ]
    }
}

 | comp=IoTAgent

```

the same expression is applied to all attributes

**Comment by AlvaroVega** [comment]
related issues https://github.com/telefonicaid/iotagent-node-lib/issues/635

**Comment by AlvaroVega** [comment]
a multi measure like
 [{"type":"string","value":7,"object_id":"cont3","name":"vol"},{"type":"string","value":9,"object_id":"cont1","name":"vol"},{"type":"string","value":8,"object_id":"cont2","name":"vol"}] 

is simplified to a context {"vol":8,"cont3":7,"cont1":9,"cont2":8} 

in https://github.com/telefonicaid/iotagent-node-lib/blob/68607d1e1f4fd3eafcaa3620590748abba218850/lib/plugins/jexlParser.js#L66-L96

**Comment by AlvaroVega** [comment]
There is a kind of workaround:
```json
{
 "devices": [
 {
 "device_id": "testmulti",
 "protocol": "IoTA-JSON",
 "expressionLanguage": "jexl",
 "attributes": [
 {
 "object_id": "vol1",
 "name": "vol",
 "type": "string",
 "expression" : "cont1*100",
 "entity_name": "WaterMeterSoria01",
 "entity_type": "WaterMeter"
 },
 {
 "object_id": "vol2",
 "name": "vol",
 "type": "string",
 "expression" : "cont2*100",
 "entity_name": "WaterMeterSoria02",
 "entity_type": "WaterMeter"
 },
 {
 "object_id": "vol3",
 "name": "vol",
 "type": "string",
 "expression" : "cont3*100",
 "entity_name": "WaterMeterSoria03",
 "entity_type": "WaterMeter"
 },
 {
  "object_id": "cont1",
  "name": "cont1",
  "type": "string"
 },
 {
 "object_id": "cont2",
  "name": "cont2",
  "type": "string"
 },
 {
 "object_id": "cont3",
  "name": "cont3",
  "type": "string"
 }
 ]
 }
 ]
}
```

then a measure like: 
$ mosquitto_pub -h localhost -t /json/apikey/testmulti/attrs -m '{"cont1": 4, "cont2": 5, "cont3":6}'

updates the following entities:

```
        "entities": [
            {
                "id": "thing:testmulti",
                "type": "thing",
                "cont1": {
                    "type": "string",
                    "value": 4,
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-24T08:21:38.707Z"
                        }
                    }
                },
                "cont2": {
                    "type": "string",
                    "value": 5,
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-24T08:21:38.707Z"
                        }
                    }
                },
                "cont3": {
                    "type": "string",
                    "value": 6,
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-24T08:21:38.707Z"
                        }
                    }
                },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-24T08:21:38.707Z"
                }
            },
            {
                "id": "WaterMeterSoria01",
                "type": "WaterMeter",
                "vol": {
                    "type": "string",
                    "value": "400",
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-24T08:21:38.707Z"
                        }
                    }
                },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-24T08:21:38.707Z"
                }
            },
            {
                "id": "WaterMeterSoria02",
                "type": "WaterMeter",
                "vol": {
                    "type": "string",
                    "value": "500",
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-24T08:21:38.707Z"
                        }
                    }
                },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-24T08:21:38.707Z"
                }
            },
            {
                "id": "WaterMeterSoria03",
                "type": "WaterMeter",
                "vol": {
                    "type": "string",
                    "value": "600",
                    "metadata": {
                        "TimeInstant": {
                            "type": "DateTime",
                            "value": "2020-11-24T08:21:38.707Z"
                        }
                    }
                },
                "TimeInstant": {
                    "type": "DateTime",
                    "value": "2020-11-24T08:21:38.707Z"
                }
            }
```

**Comment by chicco785** [comment]
the problem seems to be that the first attribute `name` with value `vol` is always selected. when multiple context with the same name are available, only the one "mapped" to entity should be used. the issue applies also to the legacy expression.

**Comment by chicco785** [comment]
we also a similar issue related to combining expressions and multientity.

when registering a device that use expression in the entity names:

```
{
"apikey": "RUtaOi9FbnZpcm9ubWVudE1hbmFnZW1lbnQvRGVtbzpEZXZpY2U6MQ==",
"entity_type": "Device",
"resource": "/iot/json",
"service": "EKZ",
"service_path": "/EnvironmentManagement/Demo",
"attributes": [
    {
        "object_id": "dateObservedFrom",
        "name": "dateObservedFrom",
        "type": "DateTime",
        "entity_name": "urn:ngsi-ld:TrafficFlowObserved:${@name}-${@laneId}-${@laneDirection}-${@vehicleType}",
        "entity_type": "TrafficFlowObserved"
    },
    {
        "object_id": "laneId",
        "name": "laneId",
        "type": "Number",
        "entity_name": "urn:ngsi-ld:TrafficFlowObserved:${@name}-${@laneId}-${@laneDirection}-${@vehicleType}",
        "entity_type": "TrafficFlowObserved"
    },
    {
        "object_id": "laneDirection",
        "name": "laneDirection",
        "type": "Text",
        "entity_name": "urn:ngsi-ld:TrafficFlowObserved:${@name}-${@laneId}-${@laneDirection}-${@vehicleType}",
        "entity_type": "TrafficFlowObserved"
    },
    {
        "object_id": "vehicleType",
        "name": "vehicleType",
        "type": "Text",
        "entity_name": "urn:ngsi-ld:TrafficFlowObserved:${@name}-${@laneId}-${@laneDirection}-${@vehicleType}",
        "entity_type": "TrafficFlowObserved"
    },
    {
        "object_id": "occupancy",
        "name": "occupancy",
        "type": "Number",
        "entity_name": "urn:ngsi-ld:TrafficFlowObserved:${@name}-${@laneId}-${@laneDirection}-${@vehicleType}",
        "entity_type": "TrafficFlowObserved"
    },
    {
        "object_id": "serialNumber",
        "name": "serialNumber",
        "type": "Text"
    },
    {
        "object_id": "deviceState",
        "name": "deviceState",
        "type": "Text"
    },
    {
        "object_id": "roadCode",
        "name": "name",
        "type": "Text",
        "entity_name": "urn:ngsi-ld:TrafficFlowObserved:${@name}-${@laneId}-${@laneDirection}-${@vehicleType}",
        "entity_type": "TrafficFlowObserved"
    }
],
"static_attributes": []
}
```

and sending the payload (json agent):

```
{
	"dateObservedFrom": 16.6,
	"laneId": 1,
    "laneDirection": "backward",
    "vehicleType": "c1",
    "occupancy": 0.65,
    "serialNumber": "sada",
    "deviceState": "ok",
    "roadCode": "uberlandstrasse002"
}
```

agents generates also empty entities that cause bad response from orion:

```
{
 "url": "http://orion-orion.prod:1026/v2/op/update",
 "method": "POST",
 "headers": {
     "fiware-service": "EKZ",
     "fiware-servicepath": "/EnvironmentManagement/Demo"
 },
 "json": {
     "actionType": "append",
     "entities": [
         {
             "id": "Device:VK0001",
             "type": "Device",
             "serialNumber": {
                 "type": "Text",
                 "value": "sada",
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "deviceState": {
                 "type": "Text",
                 "value": "ok",
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "TimeInstant": {
                 "type": "DateTime",
                 "value": "2020-11-24T11:07:59.975Z"
             }
         },
         {
             "id": "urn:ngsi-ld:TrafficFlowObserved:uberlandstrasse002-1-backward-c1",
             "type": "TrafficFlowObserved",
             "dateObservedFrom": {
                 "type": "DateTime",
                 "value": 16.6,
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "laneId": {
                 "type": "Number",
                 "value": 1,
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "laneDirection": {
                 "type": "Text",
                 "value": "backward",
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "vehicleType": {
                 "type": "Text",
                 "value": "c1",
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "occupancy": {
                 "type": "Number",
                 "value": 0.65,
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "name": {
                 "type": "Text",
                 "value": "uberlandstrasse002",
                 "metadata": {
                     "TimeInstant": {
                         "type": "DateTime",
                         "value": "2020-11-24T11:07:59.975Z"
                     }
                 }
             },
             "TimeInstant": {
                 "type": "DateTime",
                 "value": "2020-11-24T11:07:59.975Z"
             }
         },
         {
             "id": "urn:ngsi-ld:TrafficFlowObserved:uberlandstrasse002-1-backward-c1",
             "type": "TrafficFlowObserved"
         },
         {
             "id": "urn:ngsi-ld:TrafficFlowObserved:uberlandstrasse002-1-backward-c1",
             "type": "TrafficFlowObserved"
         },
         {
             "id": "urn:ngsi-ld:TrafficFlowObserved:uberlandstrasse002-1-backward-c1",
             "type": "TrafficFlowObserved"
         },
         {
             "id": "urn:ngsi-ld:TrafficFlowObserved:uberlandstrasse002-1-backward-c1",
             "type": "TrafficFlowObserved"
         },
         {
             "id": "urn:ngsi-ld:TrafficFlowObserved:uberlandstrasse002-1-backward-c1",
             "type": "TrafficFlowObserved"
         }
     ]
 }
}
```

**Comment by chicco785** [comment]
so the problem for our case is here:

https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/multiEntity.js#L261-268

`_.pluck` duplicates the entities since it's not filtering unique values.

**Comment by fgalan** [comment]
This has been automatically closed by merging PR #952. However, I prefer that @AlvaroVega could have a look first, as original reporter of the issue, to ensure everything is fine before closing.
